File: languages/rust/src/operations/quick_sort.rs
/**
 * Contains the implementation of QuickSort.
 */

pub fn partition<T: Ord>(arr: &mut [T]) -> usize {
    let len = arr.len();
    let pivot = len - 1;
    let mut store_index = 0;

    for i in 0..len - 1 {
        if arr[i] < arr[pivot] {
            arr.swap(i, store_index);
            store_index += 1;
        }
    }
    arr.swap(store_index, pivot);
    store_index
}

pub fn quick_sort<T: Ord>(arr: &mut [T]) -> &mut [T] {
    if arr.len() > 1 {
        let pivot = partition(arr);
        quick_sort(&mut arr[0..pivot]);
        quick_sort(&mut arr[pivot + 1..]);
    }
    arr
}

/*******************************************************
 * Unit Tests
 ******************************************************/

#[test]
fn test_quick_sort_zero_elements() {
    // Arrange
    let mut unsorted: Vec<i32> = Vec::new();
    let expected_result: Vec<i32> = Vec::new();

    // Act
    let actual_result = quick_sort(&mut unsorted);

    // Assert
    assert_eq!(expected_result, actual_result);
}

#[test]
fn test_quick_sort_single_element() {
    // Arrange
    let mut unsorted: [&str; 1] = ["badger"];
    let expected_result: [&str; 1] = ["badger"];

    // Act
    let actual_result = quick_sort(&mut unsorted);

    // Assert
    assert_eq!(expected_result, actual_result);
}

#[test]
fn test_quick_sort_multiple_odd_elements() {
    // Arrange
    let mut unsorted: [&str; 9] = ["zinc", "trap", "tree", "trait", "badger", "apple", "chemistry", "pig", "zoo"];
    let expected_result: [&str; 9] = ["apple", "badger", "chemistry", "pig", "trait", "trap", "tree", "zinc", "zoo"];

    // Act
    let actual_result = quick_sort(&mut unsorted);

    // Assert
    assert_eq!(expected_result, actual_result);
}

#[test]
fn test_quick_sort_multiple_even_elements() {
    // Arrange
    let mut unsorted: [&str; 10] = ["trap", "apple", "chemistry", "pig", "zoo", "tree", "trait", "badger", "zinc", "crack"];
    let expected_result: [&str; 10] = ["apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo"];

    // Act
    let actual_result = quick_sort(&mut unsorted);

    // Assert
    assert_eq!(expected_result, actual_result);
}

#[test]
fn test_quick_sort_already_sorted() {
    // Arrange
    let mut unsorted: [&str; 10] = ["apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo"];
    let expected_result: [&str; 10] = ["apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo"];

    // Act
    let actual_result = quick_sort(&mut unsorted);

    // Assert
    assert_eq!(expected_result, actual_result);
}

#[test]
fn test_quick_sort_integers() {
    // Arrange
    let mut unsorted: [i16; 9] = [4, -54, 40, 400, 2, -7, 0, 1, 4];
    let expected_result: [i16; 9] = [-54, -7, 0, 1, 2, 4, 4, 40, 400];

    // Act
    let actual_result = quick_sort(&mut unsorted);

    // Assert
    assert_eq!(expected_result, actual_result);
}

File: languages/zig/src/operations/quick_sort.zig
// Contains the implementation of QuickSort.

const std = @import("std");
const Order = std.math.Order;

inline fn swap(comptime T: type, i: usize, j: usize, arr: []T) void {
    const temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

inline fn partition(comptime T: type, arr: []T, comparator: *const fn (a: T, b: T) Order) usize {
    const len = arr.len;
    const pivot = len - 1;
    var store_index: usize = 0;

    for (0..len - 1) |i| {
        if (comparator(arr[i], arr[pivot]) == Order.lt) {
            // Swap arr[i] and arr[store_index]
            swap(T, i, store_index, arr);
            store_index += 1;
        }
    }
    // Swap arr[store_index] and arr[pivot]
    swap(T, pivot, store_index, arr);

    return store_index;
}

pub fn quickSort(comptime T: type, arr: []T, comparator: *const fn (a: T, b: T) Order) []T {
    if (arr.len > 1) {
        const pivot = partition(T, arr, comparator);
        _ = quickSort(T, arr[0..pivot], comparator);
        _ = quickSort(T, arr[pivot + 1 ..], comparator);
    }
    return arr;
}

/////////////////////////////////////////////////////////
// Unit Tests
/////////////////////////////////////////////////////////
const testing = std.testing;

const allocator = std.testing.allocator;
const StringList = std.ArrayList([]const u8);

fn compareStrings(a: []const u8, b: []const u8) std.math.Order {
    return std.mem.order(u8, a, b);
}

fn generateMutableStringArray(static_arr: [][]const u8) !StringList {
    var mutArr = StringList.init(allocator);
    errdefer mutArr.deinit();
    for (static_arr[0..]) |elem| {
        _ = try mutArr.append(elem);
    }
    return mutArr;
}

test "QuickSort: zero elements" {
    // Arrange
    var insert_arr = [_][]const u8{};
    var unsorted = try generateMutableStringArray(insert_arr[0..]);
    defer unsorted.deinit();
    const expected_results = try generateMutableStringArray(insert_arr[0..]);
    defer expected_results.deinit();

    // Act
    const actual_results = quickSort([]const u8, unsorted.items, &compareStrings);

    // Assert
    try testing.expectEqualSlices([]const u8, expected_results.items, actual_results);
}

test "QuickSort: one elements" {
    // Arrange
    var unsorted_literal = [_][]const u8{"badger"};
    var unsorted = try generateMutableStringArray(unsorted_literal[0..]);
    defer unsorted.deinit();
    var expected_results_literal = [_][]const u8{"badger"};
    const expected_results = try generateMutableStringArray(expected_results_literal[0..]);
    defer expected_results.deinit();

    // Act
    const actual_results = quickSort([]const u8, unsorted.items, &compareStrings);

    // Assert
    try testing.expectEqualSlices([]const u8, expected_results.items, actual_results);
}

test "QuickSort: odd elements" {
    // Arrange
    var unsorted_literal = [_][]const u8{ "zinc", "trap", "tree", "trait", "badger", "apple", "chemistry", "pig", "zoo" };
    var unsorted = try generateMutableStringArray(unsorted_literal[0..]);
    defer unsorted.deinit();
    var expected_results_literal = [_][]const u8{ "apple", "badger", "chemistry", "pig", "trait", "trap", "tree", "zinc", "zoo" };
    const expected_results = try generateMutableStringArray(expected_results_literal[0..]);
    defer expected_results.deinit();

    // Act
    const actual_results = quickSort([]const u8, unsorted.items, &compareStrings);

    // Assert
    try testing.expectEqualSlices([]const u8, expected_results.items, actual_results);
}

test "QuickSort: even elements" {
    // Arrange
    var unsorted_literal = [_][]const u8{ "trap", "apple", "chemistry", "pig", "zoo", "tree", "trait", "badger", "zinc", "crack" };
    var unsorted = try generateMutableStringArray(unsorted_literal[0..]);
    defer unsorted.deinit();
    var expected_results_literal = [_][]const u8{ "apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo" };
    const expected_results = try generateMutableStringArray(expected_results_literal[0..]);
    defer expected_results.deinit();

    // Act
    const actual_results = quickSort([]const u8, unsorted.items, &compareStrings);

    // Assert
    try testing.expectEqualSlices([]const u8, expected_results.items, actual_results);
}

test "QuickSort: same elements" {
    // Arrange
    var unsorted_literal = [_][]const u8{ "apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo" };
    var unsorted = try generateMutableStringArray(unsorted_literal[0..]);
    defer unsorted.deinit();
    var expected_results_literal = [_][]const u8{ "apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo" };
    const expected_results = try generateMutableStringArray(expected_results_literal[0..]);
    defer expected_results.deinit();

    // Act
    const actual_results = quickSort([]const u8, unsorted.items, &compareStrings);

    // Assert
    try testing.expectEqualSlices([]const u8, expected_results.items, actual_results);
}
